### PR TITLE
Dry layout create functionality 

### PR DESCRIPTION
**Pull request by nerudadhich**

### CODE REVIEW

File: server/handler.go [hunk 1]
```diff
@@ -257,7 +277,9 @@ func (s *Server) ApplyLayout(ctx context.Context, in *ApplyLayoutRequest) (*JobS
 	if err := in.Validate(); err != nil {
 		return nil, errors.Wrap(err, Errors_INVALID_VALUE.String())
 	}
-
+	if !in.Dry && len(in.Id) > 3 && in.Id[len(in.Id)-4:] == drysuffix{
```
Comment (Talina06) [inline server/handler.go]: endsWith := strings.HasSuffix(in.Id, "-dry") // true

File: server/handler_test.go [hunk 2]
```diff
@@ -81,9 +82,44 @@ func TestServer_SaveAndGetLayout(t *testing.T) {
...
+
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		assert.Equal(t, resp.LayoutId, layoutId+drysuffix)
 	})
 
+	t.Run("Should create a layout in the workspace with dry flag and copy state file", func(t *testing.T) {
+		//temporary saving state
```
Comment (Talina06) [inline server/handler_test.go]: 1. layout create test: will have 2 version.
2. layout create test with dry
3. check if test has 2 versions only after step 2. if > 3 then fail

**Comment by Talina06** [comment]
download metalinters @nerudadhich  for golang formatting

File: server/handler.go [hunk 3]
```diff
@@ -129,10 +132,27 @@ func (s *Server) getWorkspace(id string) (*Workspace, error) {
...
 
 // SaveLayout under the mentioned workspace ID.
-func (s *Server) SaveLayout(ctx context.Context, in *SaveLayoutRequest) (*Ok, error) {
+func (s *Server) SaveLayout(ctx context.Context, in *SaveLayoutRequest) (*SaveLayoutResponse, error) {
 	if err := in.Validate(); err != nil {
 		return nil, errors.Wrap(err, Errors_INVALID_VALUE.String())
 	}
+	layoutId := in.Id
+	if in.Dry{
+		// copy old state to dry layout or create new
```
Comment (mohandutt134) [inline server/handler.go]: add all this in a separate function

File: server/handler.go [hunk 4]
```diff
@@ -129,10 +132,27 @@ func (s *Server) getWorkspace(id string) (*Workspace, error) {
 }
 
 // SaveLayout under the mentioned workspace ID.
-func (s *Server) SaveLayout(ctx context.Context, in *SaveLayoutRequest) (*Ok, error) {
+func (s *Server) SaveLayout(ctx context.Context, in *SaveLayoutRequest) (*SaveLayoutResponse, error) {
 	if err := in.Validate(); err != nil {
 		return nil, errors.Wrap(err, Errors_INVALID_VALUE.String())
 	}
+	layoutId := in.Id
+	if in.Dry{
```
Comment (mohandutt134) [inline server/handler.go]: There are lot of formatting issues, please use gofmt or goimports by integrating it in your IDE

File: server/handler.go [hunk 5]
```diff
@@ -129,10 +132,27 @@ func (s *Server) getWorkspace(id string) (*Workspace, error) {
...
+		// copy old state to dry layout or create new
+		key := path.Join(state, in.WorkspaceId, layoutId)
+		stateValue, err := s.store.GetKey(key)
+		if err != nil {
+			return nil, err
+		}
+		layoutId = layoutId+ drysuffix
+		if len(string(stateValue)) > 0 {
+			key = path.Join(state, in.WorkspaceId, layoutId)
+			err = s.store.SaveKey(key, stateValue)
```
Comment (mohandutt134) [inline server/handler.go]: preffer following pattern in such cases
```
if err := s.store.SaveKey(key, stateValue); err != nil {
}
```

File: server/handler.go [hunk 6]
```diff
@@ -129,10 +132,27 @@ func (s *Server) getWorkspace(id string) (*Workspace, error) {
...
+			if err != nil {
+				return nil, err
+			}
+		}
```
Comment (mohandutt134) [inline server/handler.go]: There should be a `return` here. I do not think we need to go further in case of -dry.

File: server/handler.go [hunk 7]
```diff
@@ -153,14 +173,14 @@ func (s *Server) SaveLayout(ctx context.Context, in *SaveLayoutRequest) (*Ok, er
 	}
 
 	// Create layout instance to be saved for given ID and plan.
-	layout := types.Layout{Id: in.Id, Plan: p, Status: int32(Status_INACTIVE)}
+	layout := types.Layout{Id: layoutId, Plan: p, Status: int32(Status_INACTIVE)}
```
Comment (mohandutt134) [inline server/handler.go]: It should be in.Id

File: server/handler_test.go [hunk 8]
```diff
@@ -81,9 +82,75 @@ func TestServer_SaveAndGetLayout(t *testing.T) {
...
+
+	t.Run("Should create a layout in the workspace with dry flag", func(t *testing.T) {
+		req := &SaveLayoutRequest{Id: layoutId, WorkspaceId: workspaceId, Plan: pBytes, Dry:true}
+		resp, err := server.SaveLayout(context.Background(), req)
+
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		assert.Equal(t, resp.LayoutId, layoutId+drysuffix)
```
Comment (mohandutt134) [inline server/handler_test.go]: avoid calculating new layoutID again and again just save it in a variable like `newLayoutID := layoutId + drysuffix`

File: server/handler_test.go [hunk 9]
```diff
@@ -81,9 +82,75 @@ func TestServer_SaveAndGetLayout(t *testing.T) {
...
+			Id:   layoutId+drysuffix,
+			Plan: map[string]json.RawMessage{},
+		}
+
+		vAfterSave, err := store.GetVersions(&l, tree)
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		assert.Equal(t, 2, len(vAfterSave))
```
Comment (mohandutt134) [inline server/handler_test.go]: I do not think dry layouts should be versioned. I see it as unnecessary overhead.

File: server/handler_test.go [hunk 10]
```diff
@@ -81,9 +82,75 @@ func TestServer_SaveAndGetLayout(t *testing.T) {
...
+	t.Run("Should create a layout in the workspace with dry flag", func(t *testing.T) {
+		req := &SaveLayoutRequest{Id: layoutId, WorkspaceId: workspaceId, Plan: pBytes, Dry:true}
+		resp, err := server.SaveLayout(context.Background(), req)
+
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		assert.Equal(t, resp.LayoutId, layoutId+drysuffix)
+
```
Comment (mohandutt134) [inline server/handler_test.go]: func TestServer_SaveGetLayout_dry()

File: server/handler_test.go [hunk 11]
```diff
@@ -264,6 +263,108 @@ func TestServer_SaveAndGetLayout(t *testing.T) {
...
+	t.Run("Should create a layout in the workspace with dry flag", func(t *testing.T) {
+		req := &SaveLayoutRequest{Id: layoutId, WorkspaceId: workspaceId, Plan: pBytes, Dry: true}
+		resp, err := server.SaveLayout(context.Background(), req)
+
+		if err != nil {
+			t.Fatal(err)
+		}
+		newLayoutID := layoutId + drysuffix
+		assert.Equal(t, resp.LayoutId, newLayoutID)
```
Comment (mohandutt134) [inline server/handler_test.go]:
```suggestion
		assert.Equal(t, newLayoutID, resp.LayoutId)
```
`assert.Equal(t, <expected>, <got>)`

File: server/handler_test.go [hunk 12]
```diff
@@ -264,6 +263,108 @@ func TestServer_SaveAndGetLayout(t *testing.T) {
...
+			Id:   newLayoutID,
+			Plan: map[string]json.RawMessage{},
+		}
+
+		vAfterSave, err := store.GetVersions(&l, tree)
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		assert.Equal(t, 1, len(vAfterSave))
```
Comment (mohandutt134) [inline server/handler_test.go]: check for plan and id as well.

File: server/handler_test.go [hunk 13]
```diff
@@ -264,6 +263,108 @@ func TestServer_SaveAndGetLayout(t *testing.T) {
...
+
+	vBytes, err := ioutil.ReadFile("../tmpl/testdata/vars.json")
+	if err != nil {
+		t.Error(err)
+	}
+
+	pBytes, _ := json.Marshal(plan)
+	vBytes = uglyJson(vBytes)
+
+	t.Run("Should create a layout in the workspace with dry flag", func(t *testing.T) {
```
Comment (mohandutt134) [inline server/handler_test.go]:
```suggestion
	t.Run("Should create a layout in the workspace with dry flag with empty state", func(t *testing.T) {
```

File: server/handler_test.go [hunk 14]
```diff
@@ -264,6 +263,108 @@ func TestServer_SaveAndGetLayout(t *testing.T) {
...
+
+		vAfterSave, err := store.GetVersions(&l, tree)
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		assert.Equal(t, 1, len(vAfterSave))
+	})
+
+	t.Run("Should create a layout in the workspace with dry flag and copy state file", func(t *testing.T) {
```
Comment (mohandutt134) [inline server/handler_test.go]:
```suggestion
	t.Run("Should create a layout in the workspace with dry flag with existing state", func(t *testing.T) {
```

**Pull request merged**